### PR TITLE
Add cobbler inventory plugin

### PR DESCRIPTION
**Pull request by opoplawski**
##### SUMMARY

This converts the cobbler inventory script to an inventory plugin.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
cobbler

##### ADDITIONAL INFORMATION
I'm also looking into maintaining this as a separate collection as part of the cobbler project here: https://github.com/cobbler/ansible.  I've requested the cobbler namespace on galaxy, but this may be problematic as there is an existing 'cobbler' user - although with no apparent content.

### CODE REVIEW

File: plugins/inventory/cobbler.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+# Copyright (C) 2020 Orion Poplawski <orion@nwra.com>
+# Copyright (c) 2020 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: cobbler
+    plugin_type: inventory
+    short_description: cobbler inventory source
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:11]:
```suggestion
    short_description: cobbler inventory source
    version_added: 1.0.0
```

File: plugins/inventory/cobbler.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    description:
+        - Get inventory hosts from the cobbler service.
+        - "Uses a configuration file as an inventory source, it must end in ``.cobbler.yml`` or ``.cobbler.yaml`` and has a ``plugin: cobbler`` entry."
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:14]:
```suggestion
        - "Uses a configuration file as an inventory source, it must end in C(.cobbler.yml) or C(.cobbler.yaml) and has a C(plugin: cobbler) entry."
```

File: plugins/inventory/cobbler.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    extends_documentation_fragment:
+        - inventory_cache
+    options:
+      plugin:
+        description: the name of this plugin, it should always be set to 'cobbler' for this plugin to recognize it as it's own.
+        required: True
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:20]:
```suggestion
        required: true
```
We use either `true`/`false` or `yes`/`no` in YAML documentation.

File: plugins/inventory/cobbler.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    plugin_type: inventory
+    short_description: cobbler inventory source
+    description:
+        - Get inventory hosts from the cobbler service.
+        - "Uses a configuration file as an inventory source, it must end in ``.cobbler.yml`` or ``.cobbler.yaml`` and has a ``plugin: cobbler`` entry."
+    extends_documentation_fragment:
+        - inventory_cache
+    options:
+      plugin:
+        description: the name of this plugin, it should always be set to 'cobbler' for this plugin to recognize it as it's own.
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:19]:
```suggestion
        description: The name of this plugin, it should always be set to C(cobbler) for this plugin to recognize it as it's own.
```

File: plugins/inventory/cobbler.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        required: True
+        choices: ['cobbler']
+      url:
+        description: url to cobbler
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:23]:
```suggestion
        description: URL to cobbler.
```

File: plugins/inventory/cobbler.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        default: 'http://cobbler/cobbler_api'
+        env:
+            - name: COBBLER_SERVER
+      user:
+        description: cobbler authentication user
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:28]:
```suggestion
        description: Cobbler authentication user.
```
If `cobbler` needs to be written with a lower-case `c`, just add the period. (Similar changes for the other descriptions.)

File: plugins/inventory/cobbler.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        type: list
+        default: [ 'mgmt_classes', 'owners', 'status' ]
+      group:
+        description: group to place all hosts into
+        default: cobbler
+      group_prefix:
+        description: prefix to apply to cobbler groups
+        default: cobbler_
+      want_facts:
+        description: Toggle, if True the plugin will retrieve host facts from the server
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:52]:
```suggestion
        description: Toggle, if C(true) the plugin will retrieve host facts from the server
```
`C(...)` is used for values (https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation).

File: plugins/inventory/cobbler.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+from ansible.module_utils._text import to_bytes, to_native, to_text
+from ansible.module_utils.common._collections_compat import MutableMapping
+from ansible.module_utils.six import iteritems
+from ansible.plugins.inventory import BaseInventoryPlugin, Cacheable, to_safe_group_name
+
+# 3rd party imports
+try:
+    import xmlrpclib as xmlrpc_client
+except ImportError:
+    import xmlrpc.client as xmlrpc_client
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:77]: This should also be put into a `try:`/`except`. Set a variable `HAS_XMLRPC_CLIENT` to `False` if both `import`s fail (and to `True` if they succeed), and fail inside the plugin code if the variable is `False` (see other plugins).

File: plugins/inventory/cobbler.py (hunk 1 above)
Comment (Andersson007) [inline plugins/inventory/cobbler.py:11]:
```suggestion
    short_description: Cobbler inventory source
```

File: plugins/inventory/cobbler.py (hunk 3 above)
Comment (Andersson007) [reply plugins/inventory/cobbler.py:20]: we should use `yes/no` according to the doc

File: plugins/inventory/cobbler.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+        required: True
+        choices: ['cobbler']
+      url:
+        description: url to cobbler
+        default: 'http://cobbler/cobbler_api'
+        env:
+            - name: COBBLER_SERVER
+      user:
+        description: cobbler authentication user
+        required: False
```
Comment (Andersson007) [inline plugins/inventory/cobbler.py:29]:
```suggestion
        required: no
```
and so on

File: plugins/inventory/cobbler.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        env:
+            - name: COBBLER_USER
+      password:
+        description: cobbler authentication password
+        required: False
```
Comment (Andersson007) [inline plugins/inventory/cobbler.py:34]:
```suggestion
        description: cobbler authentication password
        required: no
```
@felixfontein don't we need to put `type:` as well? (i'm not sure it's relevant for plugins)

File: plugins/inventory/cobbler.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        env:
+            - name: COBBLER_PASSWORD
+      exclude_profiles:
+        description: profiles to exclude from inventory
+        type: list
+        default: []
```
Comment (Andersson007) [inline plugins/inventory/cobbler.py:40]:
```suggestion
        description: Profiles to exclude from inventory.
        type: list
        elements: str
        default: []
```
and in other places
are `elements: ` relevant here? @felixfontein

Comment (opoplawski) [reply plugins/inventory/cobbler.py]: I think so, added.

**Comment by Andersson007** [comment]
@felixfontein do you have any other suggestions / questions or we could merge this?

File: plugins/inventory/cobbler.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+        default: cobbler_
+      want_facts:
+        description: Toggle, if C(true) the plugin will retrieve host facts from the server
+        type: boolean
+        default: yes
+'''
+
+EXAMPLES = '''
+# my.cobbler.yml
+plugin: cobbler
```
Comment (felixfontein) [inline plugins/inventory/cobbler.py:65]:
```suggestion
plugin: community.general.cobbler
```

**Comment by Andersson007** [comment]
@opoplawski thanks for introducing the feature!
@felixfontein thanks for reviewing!
merged

**Comment by opoplawski** [comment]
w00t - yeah I'm using it in production and so far so good.

**Pull request merged**